DigitalLearningSolutions.Data/Services/TrackerActionService.cs

```diff
@@ -51,5 +67,34 @@ public TrackerActionService(ITutorialContentDataService tutorialContentDataServi
 
             return ccObjectives.Any() ? new TrackerObjectiveArrayCc(ccObjectives) : null;
         }
+
+        public TrackerEndpointResponse StoreDiagnosticJson(int? progressId, string? diagnosticOutcome)
+        {
+            if (!progressId.HasValue || string.IsNullOrEmpty(diagnosticOutcome))
+            {
+                return TrackerEndpointResponse.StoreDiagnosticScoreException;
+            }
+
+            var json = JsonConvert.DeserializeObject<IEnumerable<JObject>>(diagnosticOutcome);
+
+            try
+            {
+                foreach (var obj in json)
+                {
+                    var tutorialId = (int)obj["tutorialid"];

```

Comment:
Given we know what the data structure should be, if a model was defined for this such as `DiagnosticOutcome` containing the two relevant properties then the code could become:
```
var diagnosticOutcomes = JsonConvert.DeserializeObject<IEnumerable<DiagnosticOutcome>>(diagnosticOutcome);
try
{
  foreach (var diagnosticOutcome in diagnosticOutcomes)
  {
    progressService.UpdateDiagnosticScore(progressId.Value, diagnosticOutcome .TutorialId, diagnosticOutcome.MyScore);
    ....
```
We can then reduce the need for casting and using magic strings here to identify properties of the object.

Reply:
Gave this a go. This means we'll treat the keys in the json as case insensitive, is that okay? I know @SteveJacksonSoft mentioned we want to keep tracker endpoints as close to the old behavior as possible.

I'm getting some issues when testing - sometimes entering an unexpected key still gets parsed to TutorialId or MyScore, sometimes I get Result:-25, and sometimes it just stalls out. Let me know if I missed something if or we should go back to the magic string way.

Reply:
**Regarding case sensitivity:**
In `.net core` the built in JSON deserialisation is case sensitive.
As we are using `JSON.NET` library, to achieve case sensitivity then you would need to add a custom converter.
@SteveJacksonSoft is their a requirement to be this strict? Out of the box the original lower-case keys would be accepted and therfore be compatible - it just additionally allows the keys to be camel/upper/etc case too.

**Regarding unexpected keys:**
In order to handle this you will need to change the default option by passing in `JsonSerializerSettings` as the second parameter to `DeserializeObject`, e.g.:
```
 var diagnosticOutcomes = JsonConvert.DeserializeObject<IEnumerable<DiagnosticOutcome>>(
  diagnosticOutcome,
  new JsonSerializerSettings { MissingMemberHandling = MissingMemberHandling.Error }
);
```
This will cause an `exception` which will be caught higher up and ensure a `Result:-25` response.
If implementing the above it will be useful to re-instante the test cases that focus on `unexpected keys` that were removed.

Reply:
Thanks, I've added the MissingMemberHandling to handle the unexpected key, and also added required attributes on the properties of DiagnosticOutcome so that an error will occur if one of the expected keys is missing

Reply:
I think it's fine if we're case-insensitive with the json in DiagnosticOutcome..

Reply:
thanks @SteveJacksonSoft